Display app version (task #73) + mobile optimalization (bug #62) + validation (task #43)

KamilZm opened the pull request:
@zzen @michalblaha 
Vyřešil jsem to tak, že se informaci o verzi bere ze souboru `/appversion.txt`. Je to pro generování jednodušší, než něco vyměňovat v HTML souboru a také je to bezpečnější, protože nehrozí, že by se mohl index.html tím zásahem díky nějakému problému rozdrbat. 
@zzen Pošlete prosím @michalblaha info, jak zjistit hash commitu, který jde do produkce
@michalblaha Zjištěný hash prosím při publikování uložte do `/appversion.txt`
Jako výchozí je v `/appversion.txt` verze `000000100000000000000000000000000000000`, takže kdyby generování selhalo, zobrazí se ta (resp. 0000001) nebo žádná, pokud obsahem toho souboru nebude pouze SHA1).

KamilZm committed:
Display app version - task #73

KamilZm committed:
Added scolling to part of form after user press button

KamilZm commented:
V commitu 5e21455 přidáno scrollování na prvek formuláře po stisknutí tlačítka, aby se uživatel (hlavně v mobilní verzi) nedostal vždy na začátek stránky, protože to na mě působili strašně zmatečně (stiskl tlačítko a viděl zase od začátku to, co předtím).

KamilZm committed:
Better validation - ISSUE #43

KamilZm commented:
V commitu 50f1a6f kompletně předělány validace. Bohužel se mi zatím nepodařilo udělat rozumnou validaci na autocomplete PSČ - když se uživatel vrátí zpět z náhledu to blbě vypadá, protože se to tváří, že není položka vybraná, ale je, jen to autocomplete nezobrazuje, protože tam nemá žádný přiřazení objekt. Ale data jsou OK. 

Je to teď nachystané tak aby šel přidat i validační sumář. Ale úplně čisté řešení to není, lepší by bylo, kdyby se např. property v userdata nastavovaly přes setter a přímo se prováděla validace. Ale do toho jsem zatím nešel.

KamilZm committed:
Some fix in validation + restore postal code

KamilZm commented:
V commitu fc5c9a8 je jako hlavní úprava pro zobrazování focusu+chyb u na PSČ a také opravena věc, která mě od začátku vadila - když se uživatel vrátil z náhledu, tak nezůstalo vyplněné PSČ. Teď už zůstane (resp. se automaticky vyplní na pozadí)

KamilZm committed:
Refactoring
- move /assets/files/appversion.txt to root
- add config file for voting days
- refacfor form (only) for independ on language
- new validation system

KamilZm reviewed (commented):
V commitu 19519b6 jsem udělal velké změny, protože mě už nebavilo pořád paralelně udržovat  CZ a EN verzi formuláře, protože ta práce s položkami (input boxy, validace apod.) byla šílená. Teď je jen jedna, univerzální verze. Týká se to jen formuláře, ne obecně.  
Jazykově nezávislá část aplikace je ohraničena komentářem `fully independent part on selected language` tam to prosím držte a nepište dovnitř lokalizované texty.
Dále jsem udělal konfigurační soubor, kde jsou zatím data voleb.
Předělal jsem validace a už se umí zobrazit kompletní validační sumář nebo jen závažnější chyby s popisem (přepíná se to nastavuje v konfiguraci).
PS: Snažil jsem se kód komentovat, ale v EN nejsem moc silný..

zzen committed:
Merge branch 'euro2019' into euro2019

zzen commented:
Kamile, build je rozbitý protože je potřeba přidat nové soubory appconfig.js. appconfig.txt do now.json konfigurace, např. takhle:

```javascript
    "builds": [
        { "src": "index.html", "use": "@now/static" },
        { "src": "appconfig.js", "use": "@now/static" },
        { "src": "appversion.txt", "use": "@now/static" },
        { "src": "favicon.ico", "use": "@now/static" },
        { "src": "assets/**", "use": "@now/static" }
    ],
```

Když mi ten pullrequest upravíte (netřeba zavírat, stačí pushnout další commit do vašeho repa https://github.com/KamilZm/Volicsky-Prukaz/tree/euro2019/) tak já budu moct potom zkontrolovat tu výslednou deploynutou verzi.

zzen commented on index.html:
```diff
@@ -133,7 +134,7 @@
                 <h1>Volby do Evropského parlamentu 2019</h1>
                 <p>
-                    <span>Volby do Evropského parlamentu se uskuteční <strong>24.&nbsp;5.&nbsp;2019 (od 14:00 do 22:00)</strong> a <strong>25&nbsp;.5.&nbsp;2019 (od 8:00 do 14:00)</strong></span>
+                    <span>Volby do Evropského parlamentu se uskuteční ve dnech <strong>{{decoder(this.getVotingIntervalWithTime)}}</strong></span>
```
Kamile - není mi jasné, k čemu přesně slouží decoder (a &nbsp; znaky jsem nikde nenašel), nicméně výstup z tohoto formátování těch datumů je jiný než původní: `24. 5. 2019 (od 14:00 do 22:00) a 25 .5. 2019 (od 8:00 do 14:00)` vs `24.05.2019 (14:00-22:00) až 25.05.2019 (08:00-14:00)`. Doporučoval bych na to formátování dat použít ten moment.js který už v jiných místech aplikace použit je.

KamilZm replied:
decoder je funkce zajišťující, aby se v {{ }} nezobrazovaly HTML entity ale aplikovaly/renderovaly se. Bez něj by tam bylo zobrazeno tohle (nechtěl jsem, aby se to mohlo zalomit):
`24.&nbsp;5.&nbsp;2019 (14:00&nbsp;&#8209;&nbsp;22:00)` 
Moment.js - OK

zzen commented on index.html:
```diff
@@ -165,10 +166,10 @@ <h2>Chci volit v kterémkoliv volebním obvodě v ČR</h2>
                 <ul class="new-p">
-                    <li><strong class="emphasized">osobně</strong><span> – dostavte se přímo na úřad s platným dokladem totožnosti nejpozději 22. 5. 2019,</span></li>
-                    <li><strong class="emphasized">písemně</strong><span> – žádost musíte úřadu doručit do 17. 5. 2019 v 16:00 pomocí vaší datové schránky nebo poštou s úředně ověřeným podpisem.</span></li>
+                    <li><strong class="emphasized">osobně</strong><span> – dostavte se přímo na úřad s platným dokladem totožnosti nejpozději {{this.appConfig.votingCardRequestdeliveryLastTime.personal}},</span></li>
```
Kamile - to samé platí pro formátovací funkce tady - výstup je horší než co jsme tam měli předtím:

> platným dokladem totožnosti nejpozději 22. 5. 2019

vs

> platným dokladem totožnosti nejpozději 22.05.2019 16:00

(proč ten čas, nezáleží to na otvíracích hodinách úřadu? a pokud ano, měla bý tam být předložka `v`)

KamilZm replied:
OK, lze to upravit. Tedy pokud to neodrolujeme, viz [závěrečný komentář](https://github.com/HlidacStatu/Volicsky-Prukaz/pull/74#issuecomment-491419837).

zzen commented on index.html:
```diff
@@ -165,10 +166,10 @@ <h2>Chci volit v kterémkoliv volebním obvodě v ČR</h2>
-                    <li><strong class="emphasized">osobně</strong><span> – dostavte se přímo na úřad s platným dokladem totožnosti nejpozději 22. 5. 2019,</span></li>
-                    <li><strong class="emphasized">písemně</strong><span> – žádost musíte úřadu doručit do 17. 5. 2019 v 16:00 pomocí vaší datové schránky nebo poštou s úředně ověřeným podpisem.</span></li>
+                    <li><strong class="emphasized">osobně</strong><span> – dostavte se přímo na úřad s platným dokladem totožnosti nejpozději {{this.appConfig.votingCardRequestdeliveryLastTime.personal}},</span></li>
+                    <li><strong class="emphasized">písemně</strong><span> – žádost musíte úřadu doručit do {{this.appConfig.votingCardRequestdeliveryLastTime.post}}h pomocí vaší datové schránky nebo poštou s úředně ověřeným podpisem.</span></li>
```
to samé, předchozí verze:

> žádost musíte úřadu doručit do 17. 5. 2019 v 16:00

vs

> žádost musíte úřadu doručit do 17.05.2019 16:00h

(h nakonci, které navíc v předchozí řádce při uvedení času není, a tady bych ho také neuváděl)

KamilZm replied:
Viz výše.

KamilZm committed:
Add new files to now.json

KamilZm committed:
Merge branch 'euro2019' of https://github.com/KamilZm/Volicsky-Prukaz into euro2019

zzen commented on index.html:
```diff
@@ -177,12 +178,12 @@ <h2>Chci volit české kandidáty</h2>
-                    <li><strong class="emphasized">osobně</strong><span> – dostavte se přímo na úřad s platným dokladem totožnosti nejpozději 22. května 2019,</span></li>
-                    <li><strong class="emphasized">písemně</strong><span> – žádost musíte doručit do 17. 5. 2019 v 16:00 pomocí vaší datové schránky nebo poštou s úředně ověřeným podpisem.</span></li>
+                    <li><strong class="emphasized">osobně</strong><span> – dostavte se přímo na úřad s platným dokladem totožnosti nejpozději {{this.appConfig.votingCardRequestdeliveryLastTime.personal}}h,</span></li>
+                    <li><strong class="emphasized">písemně</strong><span> – žádost musíte doručit do {{this.appConfig.votingCardRequestdeliveryLastTime.post}}h pomocí vaší datové schránky nebo poštou s úředně ověřeným podpisem.</span></li>
```
ditto viz výše

KamilZm replied:
Viz výše.

zzen commented on index.html:
```diff
@@ -206,230 +207,59 @@ <h3>I want to vote where I have permanent residence in CZ</h3>
                 <ul class="new-p">
-                    <li><strong class="emphasized">in person</strong><span> – directly at the office with valid ID no later than 22 May 2019,</span></li>
-                    <li><strong class="emphasized">in writing</strong><span> – through your data box or by post with a certified signature. Your application must arrive by 17 May 2019 at 16:00.</span></li>
+                    <li><strong class="emphasized">in person</strong><span> – directly at the office with valid ID no later than {{this.appConfig.votingCardRequestdeliveryLastTime.personal}}h,</span></li>
```
Kromě připomínek k formátování výše, zde bylo formátování data v anglosaské formě (22 May 2019), což popravdě mělo být 22nd May 2019. Opět doporučuju použít moment.js který tohle umí dělat za vás.

KamilZm replied:
Viz výše.

KamilZm commented:
@zzen OK, omlouvám se a díky za vysvětlení, přidáno. 
Vrtalo mi teď hlavou, že jsem to nemusel dělat dříve, když jsem přidával soubory do `/assets/`, ale teď tam vidím `{ "src": "assets/**", "use": "@now/static" }`, čímž je to jasné.

zzen commented on index.html:
```diff
@@ -206,230 +207,59 @@ <h3>I want to vote where I have permanent residence in CZ</h3>
-                            Pokud chcete mít 100% jistotu, můžete se při vytváření žádosti odpojit od internetu
-                            nebo si stáhněte a vyplňte <a href="assets/files/Zadost-Volicsky-Prukaz-EU2019-CR.pdf">šablonu žádosti</a> ručně sami.
-                        </p>
+                            {{texts['privacyInfo'][this.getLanguage()]}} <a href="assets/files/Zadost-Volicsky-Prukaz-EU2019-CR.pdf">{{texts['applicationTemplateDownloadLink'][this.getLanguage()]}}</a>.
```
Kamile, já to původně také dělal formou textových lokalizací, ale pak jsem se po úvaze rozhodl to takhle nedělat, protože mi přijde, že zápory převažují nad klady. Je sice pravda, že při udržování aplikace je občas potřeba změny aplikovat dvakrát, nicméně zase je ten HTML blok relativně kompaktní a je relativně přehledné, jak se zobrazuje.

Pěkný příklad je v téhle ukázce. Původní znění bylo relativně uživatelsky přívětivé:

> Pokud chcete mít 100% jistotu, můžete se při vytváření žádosti odpojit od internetu nebo si stáhněte a vyplňte [šablonu žádosti]() ručně sami.

vs 

> Pokud chcete mít 100% jistotu, můžete se při vytváření žádosti odpojit od internetu nebo si žádost stáhněte a vyplňte ručně sami [(stažení šablony žádosti)]().

KamilZm replied:
Viz výše.

zzen commented on index.html:
```diff
@@ -438,27 +268,29 @@ <h2>My identity</h2>
-                                    max-result-items="15" more-results-exists-text="(...be more specific...)" ref="autocompleteZIP"
-                                    @selected="zipSelected">
+                                    search-from-beginning="true"  min-legth-for-search-begin="0"
+                                    max-result-items="15" more-results-exists-text="(...upřesněte zadání...)" ref="autocompleteZIP"
```
lady máme příklad nezlokalizovaného textu který z české verze prolíná do anglické protože nemá překladovou funkci

KamilZm replied:
Omlouvám se a děkuji za upozornění, přehlédl jsem to.

zzen commented on index.html:
```diff
@@ -779,54 +634,232 @@ <h2>3. We wish you good voting!</h2>
         });
+
+        // generic object for user data property
+        function getDefaultDataObject(nullable, valid, caption, validator) { 
```
Kamile - tohle mi přijde také obrovská změna kterou úplně nechápu - skoro celá aplikace je předělaná, kdekoliv je `userData` použito, tak tam na konci přibylo `.value`. Pak je yde `getDefaultDataObject()` a `getUserData()` - k čemu je celá tahle ceremonie dobrá?

KamilZm replied:
Možná je to jen tím, že neumím Javascript. 

Z původních "obyčejných" property (fullName) se skalárními hodnotami jsem udělal objekty, které mají své vlastnosti (caption, validátor atd.).  a jedna z vlastností je i jeho vlastní hodnota (value). Pokud to jde udělat tak, aby `userData.fullName` obsahovalo skalární hodnotu (to mé dnešní value) a současně mohlo existovat např. `userData.fullName.caption` a metody (`userData.fullName.validate()` atd.), rád to předělám. 

Objekt pro jednu property (fullName, city, ...) vrací funkce `getDefaultDataObject`. Funkce `getDefaultUserData` pak vrací kompletní objekt pro držení zadaných informací, kde jednotlivé property tvoří právě getDefaultDataObject.

Do funkce `getDefaultUserData` jsem to vyčlenil z důvodu, že se používá na dvou místech (druhé využití je pro tlačítko vyčištění formuláře, které často používám). Pokud bychom tlačítko na čištění vyhodili (stejně se Vám také moc nepozdávalo), lze vyhodit i `getDefaultUserData`.

zzen commented on index.html:
```diff
@@ -779,54 +634,232 @@ <h2>3. We wish you good voting!</h2>
+                    return textArea.value;
+                },
+                getLanguage: function () {  // get current language base on route path
+                    return this.$route.path == '/en/form' ? 'EN' : 'CZ';
```
jak tohle funguje? máme i `/en/preview` - to lokalizováno nebylo?

KamilZm replied:
Ne, zatím jsem lokalizoval jen část, která je na změny nejcitlivější na změny, protože dokud jsem to neměl lokalizované, tak jsem dělal strašně moc copy&paste chyb. 
Ty další části mají i jiný charakter - jsou to většinou jen texty s interpolací proměnných a tam jsem pro lokalizaci neviděl už tak velké výhody a lokalizovaná verze by vlastně vypadala +- stejně (opět jen text s proměnnými). 
Obecně se tomu nebráním, ale nemělo to pro mě prioritu, protože jsem do toho prakticky nesahal (až na přidání .value) a nevím, zda to má u těchto bloků smysl a čekal jsem, že se domluvíme.

KamilZm committed:
Replace dash in date/time texts with non-breaking hyphen

zzen reviewed (changes requested):
Kamile - pardon za neodeslání review… Poznámky jsou u jednotlivých řádek kódu ale ještě to v rychlosti shrnu:

1) je to opravdu velký pull-request a spojuje tam několik separátních vylepšení: nové validace, appversion/appconfig infrastrukturu a sloučení en/cz verze formuláře. Pro příště bych doporučoval posílat jako separátní pull-requesty - pak se nestane, že více komplikované/kontroverzní změny jako třeba ty validace+refaktoring jazykové mutace nezablokují malé a triviální změny co už dávno mohly být nasazené - jako třeba ten appversion.txt

2) z appconfig.js mám fakt radost, tak jsem to také chtěl udělat (mít explicitní konfiguraci těch variabilních dat) aby se to hezky dalo udržovat dále do budoucna. Ale vadí mi tam, že jsme snižili uživatelskou přívětivost aplikace a tak mám komentáře k formátování těch dat. Je tam už velmi flexibilní knihovna moment.js používaná na formátování datumů/času na jiných místech, doporučuju vyřešit tímto

3) z validací jsem trochu rozpačitý, protože v kombinaci s tím sloučením EN/CZ verze do lokalizovaných textu celou záležitost dost výrazně zkomplikovaly. Navíc je to nekonzistentní (sloučené jsou jen ty dvě formulářové stránky, /preview je stále ještě duální jestli na to správně koukám). Z pohledu code-review mimo jiné hodnotím i jak je jednoduché ten kód přečíst/pochopit pro někoho "kdo přijde po nás" a tady jsme myslím refaktorem dost utrpěli.

Tolik ke shrnutí - detaily viz konkrétní řádky.

KamilZm commented:
Ke shrnutí: 

Úvodem - nepovažuji to **samozřejmě** za celý refaktor a konečné řešení. Dělal jsem to, co mě nejvíc pálilo, na co jsem sahal kvůli validacím a na co byl čas.

ad 1 (pull request) - souhlas, chtěl jsem to, ale problém byl, že jsem nepřišel na to, jak udělat další PR, protože dokud jsem měl otevřený jeden, tak jsem neuměl udělat druhý. Commity se automaticky přidávaly k tomu otevřenému.

ad 2) (formáty dat) - asi není větší problém opravit. Zatím jsem tam moc neřešil EN verzi (vzhledem k času), protože tu **současnou** stejně skoro nikdo používat nebude, viz diskuze ve [Slacku](https://cesko-digital.slack.com/archives/CG2U4701K/p1557227639006400). Ten, kdo ji použije, tak to snad nějak pochopí. Do budoucna jsem to plánoval předělat.

ad zbytek (validace, částečná lokalizace a předělání 'userData' ze skalárů na objekty s logikou): 
Částečnou lokalizaci jsem ze svého pohledu vysvětlil výše. 

Tahle trojice problémů tvoří 95% tohoto PR. Jak jsem pochopil, tak mé řešení validací se Vám moc nezdá (nepřehledné) , lokalizaci nepreferujete (z Vašeho pohledu přijde výhodnější mít dvě paralelní verze) a změnu v userData zřejmě také ne. Z tohoto pohledu mi teď přijde nejjednodušší to celé, až na konfiguraci 79e260a a scrollování 5e21455 odrolovat, nebo PR nemergovat a zavřít, aby se nemusel mergovat s tím, co dělal @woodarius  a ty první dvě změny udělám znovu. 

Jak jsem již psal, dělal jsem to na své triko a zlobit se nebudu. Vy jste šéf, máte v hlavě nějakou vizi tohoto projektu a jste znalostmi asi tak 1000x přede mnou.

KamilZm committed:
Fixes from code review + another modification and (maybe)improvement
- format date and time base on language
- localise election name in header
- some texts corrections
- new properties in config file

KamilZm reviewed (commented):
Commit 153c911e obsahuje převážně úpravy z code review:
- formátování datových a časových položek, včetně národnostních verzí. 
- zlepšení některých textů
- oprava lokalizace autocomplete
Dále byly do konfigurace přidány nové položky (název voleb + další data) a byl lokalizován název, který se zobrazuje v hlavičce 

Stále za mě platí od "ad zbytek" do konce [předchozího komentáře](https://github.com/HlidacStatu/Volicsky-Prukaz/pull/74#issuecomment-491419837).

KamilZm committed:
Fix bug in validations + some optimization

KamilZm reviewed (commented):
Na řádcích 338-402 je pouze opraveno odsazení

KamilZm commented:
PR po dohodě s @woodarius uzavírám. Provedené úpravy budou znovu vloženy přímo do repo do zvláštní větve (ne přes fork).
